File: ui/packages/desktop/src/main/bindings/bindings.ts
import * as lonnanote from 'lonanote-core-node';

export const init = () => {
  const error = lonnanote.init();
  if (error != null) {
    throw new Error(error);
  }
};

export const invoke = (
  key: string,
  args?: string | null | undefined,
): string | null | undefined => {
  const res = lonnanote.invoke(key, args);
  return res;
};

export const getCommandKeys = (): string[] => {
  return lonnanote.getCommandKeys();
};

export const getCommandLen = (): number => {
  return Number(lonnanote.getCommandLen());
};

export const invokeAsync = async (
  key: string,
  args?: string | null | undefined,
): Promise<string | null | undefined> => {
  const res = await lonnanote.invokeAsync(key, args);
  return res;
};

export const getCommandAsyncKeys = (): string[] => {
  return lonnanote.getCommandAsyncKeys();
};

export const getCommandAsyncLen = (): number => {
  return Number(lonnanote.getCommandAsyncLen());
};

export const regJsFunction = (
  key: string,
  callback: (args: string | null | undefined) => Promise<string | null | undefined>,
): void => {
  lonnanote.regJsFunction(key, callback);
};

export const unregJsFunction = (key: string): void => {
  lonnanote.unregJsFunction(key);
};

export const clearJsFunction = (): void => {
  lonnanote.clearJsFunction();
};

export const getCommandJsKeys = (): string[] => {
  return lonnanote.getCommandJsKeys();
};

export const getCommandJsLen = (): number => {
  return Number(lonnanote.getCommandJsLen());
};

File: ui/flutter/lib/src/bindings/bindings.dart
import 'dart:async';
import 'dart:convert';

import 'package:lonanote/src/bindings/api/path/path.dart';
import 'package:lonanote/src/rust/api/command.dart' as command;

class Bindings {
  static bool isInited = false;

  static Future<void> _initApi() async {
    await RustPath.initPath();
  }

  static Future<void> init() async {
    if (isInited) return;
    final error = command.init();
    if (error != null) {
      throw Exception(error);
    }
    await _initApi();
    isInited = true;
  }

  static dynamic invoke({required String key, Object? value}) {
    var args = jsonEncode(value);
    var ret = command.invoke(key: key, args: args);
    if (ret == null) {
      return null;
    }
    return jsonDecode(ret);
  }

  static Future<dynamic> invokeAsync(
      {required String key, Object? value}) async {
    var args = jsonEncode(value);
    var ret = await command.invokeAsync(key: key, args: args);
    if (ret == null) {
      return null;
    }
    return jsonDecode(ret);
  }

  static BigInt getCommandLen() {
    return command.getCommandLen();
  }

  static BigInt getCommandAsyncLen() {
    return command.getCommandAsyncLen();
  }

  static List<String> getCommandKeys() {
    return command.getCommandKeys();
  }

  static List<String> getCommandAsyncKeys() {
    return command.getCommandAsyncKeys();
  }

  static void regDartFunction(
      {required String key,
      required FutureOr<String?> Function(String?) callback}) {
    command.regDartFunction(key: key, callback: callback);
  }

  static void unregDartFunction({required String key}) {
    command.unregDartFunction(key: key);
  }

  static void clearDartFunction() {
    command.clearDartFunction();
  }

  static List<String> getCommandDartKeys() {
    return command.getCommandDartKeys();
  }

  static BigInt getCommandDartLen() {
    return command.getCommandDartLen();
  }
}

void testBindings() {
  // RustApp.test();
  // RustFs.test();
  // RustPath.test();
  // RustSettings.test();
  // RustWorkspace.test();
  // RustWorkspaceManager.test();
}
